### ISSUE TITLE
Invalid multiple algorithm logic

### ISSUE THREAD

**Issue by ereOn** [Keats/jsonwebtoken#297]
The `Validation` takes a `Vec` of `Algorithm` but the validation code [here](https://github.com/Keats/jsonwebtoken/blob/147d24170497a5aa973c74ac4566e8baab97770c/src/decoding.rs#L201-205) incorrectly ensures that the JWT's algorithm is compatible with **all** the specified algorithm, instead of just ensuring it finds at least one that matches.

As it stands, it is impossible to use more than one algorithm in a validation because of this bug.

**Comment by Keats** [comment]
> As it stands, it is impossible to use more than one algorithm in a validation because of this bug.

Yeah that's intentional. Ideally we would solve it at the type level to only allow a single family but I didn't want to change the types too much.

**Comment by Keats** [comment]
Most of the time, the `Validation` struct can be defined in a const way (well lazy_static or once_cell).

> Is it possible the Validation struct gained a responsibility to ensure a specific DecodingKey is appropriate rather than the DecodingKey's family being used for this?

As in passing the decoding key to the validation? It's possible but you lose the build once only approach

**Comment by PeteX** [comment]
Yes I've got the same use case as @beanow-at-crabnebula. If the current behaviour is intentional, can I suggest updating the documentation? It currently implies that it's enough for the token to match a single algorithm from the list:

> The validation will check that the `alg` of the header is contained in the ones provided and will error otherwise.

**Comment by Keats** [comment]
Honestly IMO we probably shouldn't even have the algorithm in the Validation at all since we have a DecodingKey which has the AlgorithmFamily baked in. You might still want to enforce a specific algorithm in a family though.

**Comment by ytimenkov** [comment]
> The original bug was that you can construct a Validation struct that never passes validation, by mixing families. In my opinion a very common use-case when you're using JWKS to dynamically load in public keys with an allowlist of algorithms.

Exactly my use-case - for my dev environment I want to use public asymmetric keys from OIDC and a symmetric key. This just can't possibly pass because one family is RSA and another is HMAC.

I get that you don't want to allow `none` algorithm passed to 3rd-party validator, but if I found proper key by id I see no reason why this extra check is needed.

**Comment by arckoor** [comment]
`Validation` may be redesigned in the future, see #520 for some discussion on it

**Comment by p-lindberg** [comment]
Just chiming in to say that I also hit this problem when trying to validate JWTs using public keys from a JWKS endpoint. Seems like a common use case. Will probably resort to creating one `Validation` for each algorithm for now, as I don't want to allocate a new one on each call. The docs are very misleading here, since they imply that it just checks that one of the specified algorithms are used. Maybe worth prioritising clarifying the docs at least, to avoid more confusion?

Also, hi @ytimenkov
